src/stackmap_checker/call_stack_state.c

```diff
@@ -0,0 +1,218 @@
+#include "call_stack_state.h"
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <err.h>
+
+#define MAX_BUF_SIZE 128
+
+call_stack_state_t* get_call_stack_state(unw_cursor_t cursor,
+                                         unw_context_t context)
+{
+    uint64_t *ret_addrs = calloc(MAX_CALL_STACK_DEPTH, sizeof(uint64_t));
+    unw_word_t *bps = calloc(MAX_CALL_STACK_DEPTH, sizeof(unw_word_t));
+    unw_word_t **registers = calloc(MAX_CALL_STACK_DEPTH, sizeof(unw_word_t *));
+    uint32_t frame = 0;
+    while (unw_step(&cursor) > 0) {
+        unw_word_t off, pc;
+        unw_get_reg(&cursor, UNW_REG_IP, &pc);
+        if (!pc) {
+            break;
+        }
+        char fun_name[MAX_BUF_SIZE];
+        unw_get_proc_name(&cursor, fun_name, sizeof(fun_name), &off);
+        // Stop when main is reached.
+        if (!strcmp(fun_name, "main")) {
+            break;
+        }
+        if (!registers[frame]) {
+            // 16 registers are saved for each frame
+            registers[frame] = calloc(16, sizeof(unw_word_t));
+        }
+        unw_get_reg(&cursor, UNW_X86_64_RAX, &registers[frame][0]);
+        unw_get_reg(&cursor, UNW_X86_64_RDX, &registers[frame][1]);
+        unw_get_reg(&cursor, UNW_X86_64_RCX, &registers[frame][2]);
+        unw_get_reg(&cursor, UNW_X86_64_RBX, &registers[frame][3]);
+        unw_get_reg(&cursor, UNW_X86_64_RSI, &registers[frame][4]);
+        unw_get_reg(&cursor, UNW_X86_64_RDI, &registers[frame][5]);
+        unw_get_reg(&cursor, UNW_X86_64_RBP, &registers[frame][6]);
+        unw_get_reg(&cursor, UNW_X86_64_RSP, &registers[frame][7]);
+        unw_get_reg(&cursor, UNW_X86_64_R8,  &registers[frame][8]);
+        unw_get_reg(&cursor, UNW_X86_64_R9,  &registers[frame][9]);
+        unw_get_reg(&cursor, UNW_X86_64_R10, &registers[frame][10]);
+        unw_get_reg(&cursor, UNW_X86_64_R11, &registers[frame][11]);
+        unw_get_reg(&cursor, UNW_X86_64_R12, &registers[frame][12]);
+        unw_get_reg(&cursor, UNW_X86_64_R13, &registers[frame][13]);
+        unw_get_reg(&cursor, UNW_X86_64_R14, &registers[frame][14]);
+        unw_get_reg(&cursor, UNW_X86_64_R15, &registers[frame][15]);
+        // Find address of the return address.
+        ret_addrs[frame] =
+            (uint64_t)(registers[frame][UNW_X86_64_RBP] + 8);
+        // Store the current BP.
+        *(bps + frame) = registers[frame][UNW_X86_64_RBP];
+        frame++;
+    }
+    call_stack_state_t *state = malloc(sizeof(call_stack_state_t));
+    state->ret_addrs = ret_addrs;
+    state->bps       = bps;
+    state->registers = registers;
+    state->depth     = frame;
+    state->records   = NULL;
+    return state;
+}
+
+void collect_map_records(call_stack_state_t *state, stack_map_t *sm)
```

Comment:
Similarly, let's comment up all medium to large functions so we know what they do 6 months down the line.